pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,53 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.DeferredHttpAction;
+import org.pac4j.core.exception.DeferredHttpActionCallback;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 1.9.1
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw DeferredHttpAction.deferredHttpAction("Kerberos Header not found", new DeferredHttpActionCallback() {

```

Comment:
I understand the difference but my use case requires some flexibility. Here is what I have:

1) The security providers are registered via configuration
2) If the kerberos provider is registered it's available both on the indirect flow and the direct flow and I can optionally disable it on a particular URL. If I display the login page I display a kerberos link which if you click, the negotiation will occur. If I skip the login entirely because it's it only provider then also the negotiation will occur
3) On the direct flow if there are multiple providers I attempt an authentication on each and allow the successful authentication through. If I can't get a valid authentication and I determine that there is only 1 direct provider then I MAY want to do the negotiation since basic and kerberos is capable of this. I'd like to leave this option available rather than just throw an exception. So the question is how do you allow for this flexibility? I need a way to know that the header was not extracted and hence allow the caller to decide if to actually run the client specific code for the exception. Setting the headers right away don't leave any flexibility.